### ISSUE TITLE
Fix process tree state after initialization

### ISSUE THREAD

**Issue by yanivagman** [aquasecurity/tracee#1374]
Some processes may have been created/exited between the time we initialize the process tree to the time we start monitoring sched_process_xxx tracepoints. This mismatch should probably be resolved by reiterating procfs after bpf program started, removing/adding processes as required (we may even need to do this more than once, to the point when there are no more processes to add/remove).

**Comment by rafaeldtinoco** [comment]
@yanivagman 

I have been playing with this and I realized that:

- process_context_map is updated at EVERY **sched_process_fork** and **sched_process_exit**, for every pid, even the ones not being traced, as the updates occur before should_trace(), and the map is updated for execve()'s only if task is being traced.

- process_context_map does not contain already started processes, they are in userland ProcInfo structure that reads procfs during startup.

- net_events gets process contexts either from userland map (if it was already started OR created by the SchedProcess{Fork,Exec}EventID events) OR from the bpf process_context_map.

Nevertheless, 

- the userland ProcInfo structure is updated only when events are submitted to userland (and that only happens for traced processes, not for all existing ones): when event is SchedProcessForkEventID, SchedProcessExecEventID OR SchedProcessExitEventID. 

If a task was read from procfs and it is not being traced, it won't ever be removed, or updated, from the userland ProcInfo map:

1. likely a price to pay for not knowing which tasks were to be traced when starting userland code.
2. it may be removed/updated when the PID # is re-used (because of recycling pid #s)

I have played with:

1. created a bg task that will re-read procfs after loading the bpf programs, so the userland map contains processes created in between "userland code reading procfs for the first time" AND the bpf programs being loaded.

2. creating a maintenance channel that is meant for bpf code to "inform" events that are not meant to be processed or printed by events channel, like a janitor channel with smaller events. This would make userland to always have an updated map of all existing processes (for traced and non traced events).

Which one you prefer ?

(1) would still have the issue about existing processes in userland map not being updated

(2) would have correct processes list in userland, all the time, but it would require another channel for smaller janitor events (we could use multiple event types for other types of janitor work as well).

**Comment by yanivagman** [comment]
> * process_context_map is updated at EVERY **sched_process_fork** and **sched_process_exit**, for every pid, even the ones not being traced, as the updates occur before should_trace(), and the map is updated for execve()'s only if task is being traced.

We should probably fix this. No reason to update for EVERY process in `sched_process_fork` and `sched_process_exit` but not in `sched_process_exec`. Whatever we will eventually choose (info for all pids vs traced pids) should be consistent for all those three.

> * process_context_map does not contain already started processes, they are in userland ProcInfo structure that reads procfs during startup.

Yes, the idea was to "fill the gap" using the bpf map in case of a race condition (e.g. a packet received for a process before sched_process_exec/fork event was received). TBH I'm not sure this bpf map is really needed. The probability that we will find the process in the bpf map and not in the userspace map is very small. We should think if the benefit of having such a map really worths maintining it.

> (2) would have correct processes list in userland, all the time, but it would require another channel for smaller janitor events (we could use multiple event types for other types of janitor work as well).

There are benefits, but also some drawbacks for having the full process tree information in userspace. Currently, procinfo only serves net packet events, but in the future, we might want to expand its usage. One example is the new process tree we are planning to have in tracee-rules (#1330) - For sure it will need to get the initial proc status, but after that it can just be updated through sched_process_* events, so we won't really need to hold the full process tree in tracee-ebpf. Another thing to consider is the resources required to maintain the full map. Say that the user only choose to trace `ls` commands using the comm filter, do we really need to hold the full proc tree in that case? The more important use-case, however, is when tracee-ebpf is being used in conjunction with tracee-rules for runtime security detections. In that case, we do want to be sure that we have the full process tree of the whole system, as any "hole" in the process tree might be a misdetection or missing information in the signature findings.

So, my thoughts about your idea of having a "janitor" channel are still unequivocal (by the way, what do you mean by saying "channel" - is it another perf buffer? If so, we also need to take event ordering between different perf buffers into consideration). I think that if the price of having the full process tree information in userspace is not so big (need to define what that means), then we can go for it.

**Comment by rafaeldtinoco** [comment]
Yes the channel I mentioned is a channel from a new perfbuffer, the one I called janitor perf buffer. I have something coded already, will improve it and suggest do you can review.

**Comment by rafaeldtinoco** [comment]
FYIO, I'm working in this for the next days. Current implementation missed status file format for container PIDs and TIDs (and did not consider nested containers). There are other minor stuff which I'm addressing as well.

**Comment by rafaeldtinoco** [comment]
@mtcherni95 be aware of this, please. there might be some issues in current procinfo in-memory data, which serves to enrich network capture events.

**Comment by yanivagman** [comment]
Procinfo has been removed.
The above comments should be considered as part of its reimplementation in #2586
